File: rts/lua/tests/test_prim.lua
-- `main prim` 이 골든을 바이트 단위로 재현하는가 (SPEC §24).
--
--    이 한 시험이 엔진 전체와 **독립 참조 구현**(tools/gen_prim.py)의 대조다.
--    앞선 시험들이 모듈을 따로 확인했다면, 이것은 열넷 절을 한꺼번에 맞춘다.

local H = require('tests.harness')
local MAIN = require('rts.main')

H.title('prim')

local prim = MAIN.cmd_prim()
local got = H.lines(prim)
local want = H.lines(H.golden('prim.txt'))

H.check('줄 수', got.n, want.n)
local bad = 0
local first = -1
local lim = got.n < want.n and got.n or want.n
for k = 0, lim - 1 do
    if got[k] ~= want[k] then
        bad = bad + 1
        if first < 0 then
            first = k
            H.note('%d행 기대 %s', k + 1, H.repr(want[k]))
            H.note('     실제 %s', H.repr(got[k]))
        end
    end
end
H.check(string.format('%d행 전부 일치', want.n), bad, 0)

local secs = {n = 0}
for k = 0, want.n - 1 do
    if want[k]:sub(1, 3) == '== ' then
        secs[secs.n] = want[k]
        secs.n = secs.n + 1
    end
end
H.check('절 구분은 14개', secs.n, 14)
local badend = {n = 0}
for k = 0, secs.n - 1 do
    if secs[k]:sub(-3) ~= ' ==' then
        badend[badend.n] = secs[k]
        badend.n = badend.n + 1
    end
end
H.check('절 표시 형식', badend, {n = 0})
H.check('덱 지시자가 자를 수 있는 형태', secs[0], '== 1. 거리 척도 ==')
H.check('출력은 줄바꿈으로 끝난다', prim:sub(-1), '\n')

-- 절 하나만 바뀌어도 잡히는가 — 시험 자체의 민감도 확인
local uniq = {}
local nuniq = 0
for k = 0, secs.n - 1 do
    if not uniq[secs[k]] then uniq[secs[k]] = true; nuniq = nuniq + 1 end
end
H.check_true('절마다 내용이 다르다', nuniq == 14)

return H.done()

File: isorpg/ts/tests/test_prim.ts
// 프리미티브 보고서가 골든과 바이트 단위로 같은가.
import * as H from './harness';
import { primReport } from '../src/main';

export function run(): number {
  H.title('prim');

  const got = primReport();
  const want = H.golden('prim.txt');

  const gl = got.split('\n');
  const wl = want.split('\n');
  H.check('줄 수', gl.length, wl.length);
  let bad = 0;
  const n = Math.min(gl.length, wl.length);
  for (let i = 0; i < n; i++) {
    if (gl[i] !== wl[i]) {
      bad += 1;
      if (bad <= 5) {
        console.log('  ' + (i + 1) + '줄 다름');
        console.log('    기대 ' + H.show(wl[i]));
        console.log('    실제 ' + H.show(gl[i]));
      }
    }
  }
  H.check('다른 줄', bad, 0);
  H.check('전체 바이트', got === want, true);
  H.note('보고서 ' + gl.length + '줄 · ' + Buffer.byteLength(got, 'utf8') + '바이트');

  return H.done();
}
